scipy/stats/_stats.pyx

```diff
@@ -513,6 +513,101 @@ cdef double _geninvgauss_pdf(double x, void *user_data) nogil except *:
     return math.exp(_geninvgauss_logpdf_kernel(x, p, b))
 
 
+cdef double _phi(double z) nogil:
+    """evaluates the normal PDF. Used in `studentized range`"""
+    cdef double inv_sqrt_2pi = 0.3989422804014327
+    return inv_sqrt_2pi * math.exp(-0.5 * z * z)
+
+
+cdef double _Phi(double z) nogil:
+    """evaluates the normal CDF. Used in `studentized range`"""
+    return cs.ndtr(z)
+
+cdef double _genstudentized_range_cdf(int n, double[2] x, void *user_data) nogil:
+    # evaluates the integrand of Equation (3) by Batista, et al [2]
+    # destined to be used in a LowLevelCallable
+    q = (<double *>user_data)[0]
+    k = (<double *>user_data)[1]
+    df = (<double *>user_data)[2]
+
+    s = x[1]
+    z = x[0]
+
+    # harcoded constants to avoid useless log evals
+    cdef double log_2 = 0.6931471805599453
+    cdef double log_inv_sqrt_2pi = -0.9189385332046727
+
+    # terms are evaluated within logarithms when possible to avoid overflows
+    log_terms = (math.log(k) + (df / 2) * math.log(df)
+                 - (math.lgamma(df / 2) + (df / 2 - 1) * log_2)
+                 + (df - 1) * math.log(s) - (df * s * s / 2)
+                 + log_inv_sqrt_2pi - 0.5 * z * z # Normal PDF
+                 + (k - 1) * math.log(_Phi(z + q * s) - _Phi(z)))  # phi estimation.
+
+    # The remaining math is excluded from the log because it can be 0
+    return math.exp(log_terms)
+
+
+cdef double _genstudentized_range_cdf_asymptotic(double z, void *user_data) nogil:
+    # evaluates the integrand of equation (2) by Lund, Lund, page 205. [4]
+    # destined to be used in a LowLevelCallable
+    q = (<double *>user_data)[0]
+    k = (<double *>user_data)[1]
+
+    return k * _phi(z) * math.pow(_Phi(z + q) - _Phi(z), k - 1)
+
+
+cdef double _genstudentized_range_pdf(int n, double[2] x, void *user_data) nogil:
```

Comment:
Yes, `n` is the length of `x`, where `x` is an array used to pass in the variables of integration. In the case of the PDF, because there are 2 variables of integration, `n` should always be 2. It's not actually used, but It's passed in from moment for consistency's sake.